File: src/systems/BuildingEconomyAdapter.js
/**
 * Converts renderer/world building objects into the stable, renderer-agnostic
 * contract consumed by EconomySystem. Keeping this translation in one place
 * prevents the editor and the authored skyline from drifting apart.
 */

function finiteNumber(value, fallback = 0) {
  const number = Number(value);
  return Number.isFinite(number) ? number : fallback;
}

function nonNegative(value, fallback = 0) {
  return Math.max(0, finiteNumber(value, fallback));
}

function nonNegativeInteger(value, fallback = 0) {
  return Math.max(0, Math.round(finiteNumber(value, fallback)));
}

function getWorldPosition(building) {
  const source = building?.plot || building?.group?.position;
  if (!Number.isFinite(source?.x) || !Number.isFinite(source?.z)) return undefined;
  return { x: source.x, z: source.z };
}

function getFinancialRates(building, spec, { fallbackIncomePerMinute }) {
  const authoredNet = finiteNumber(
    spec.incomePerMinute,
    fallbackIncomePerMinute
  );
  const explicitRevenue = spec.revenuePerMinute ?? spec.grossIncomePerMinute;
  const explicitUpkeep = spec.upkeepPerMinute ?? spec.operatingCostPerMinute;

  return {
    grossIncomeRate: nonNegative(
      explicitRevenue ?? Math.max(0, authoredNet)
    ) / 60,
    operatingCostRate: nonNegative(
      explicitUpkeep ?? Math.max(0, -authoredNet)
    ) / 60
  };
}

const DEFAULT_SERVICE_REACH = Object.freeze({
  power: 180,
  water: 160,
  fire: 140
});

function serviceReach(spec, service, capacity) {
  if (capacity <= 0) return 0;
  return nonNegative(
    spec[`${service}Reach`] ?? spec[`${service}Radius`] ?? spec.serviceReach,
    DEFAULT_SERVICE_REACH[service]
  );
}

export function createBuildingEconomyRecord(building, {
  spec = building?.spec || {},
  id,
  fallbackIncomePerMinute = 1_200,
  fallbackEmployees,
  fallbackValue
} = {}) {
  if (!building || typeof building !== 'object') {
    throw new TypeError('building must be an object');
  }

  const recordId = id || building.economyId || building.id;
  if (typeof recordId !== 'string' || recordId.trim() === '') {
    throw new TypeError('an economy building id is required');
  }

  const amenityRadius = nonNegative(spec.amenityRadius);
  const happinessModifier = finiteNumber(
    spec.happiness ?? spec.happinessModifier,
    0
  );
  const population = nonNegativeInteger(
    spec.residents ?? spec.population ?? building.residents
  );
  const employees = nonNegativeInteger(
    spec.employees ?? building.employees,
    finiteNumber(fallbackEmployees, 0)
  );
  const { grossIncomeRate, operatingCostRate } = getFinancialRates(
    building,
    spec,
    { fallbackIncomePerMinute }
  );

  return {
    id: recordId.trim(),
    name: building.name || spec.name || recordId,
    kind: spec.category || spec.generatorType || building.businessType || building.type || 'COMMERCIAL',
    value: nonNegative(
      spec.value ?? spec.cost ?? building.value,
      finiteNumber(fallbackValue, 0)
    ),
    employees,
    jobCapacity: nonNegativeInteger(spec.jobCapacity ?? spec.jobs, employees),
    population,
    housingCapacity: nonNegativeInteger(spec.housingCapacity, population),
    status: building.status || spec.status || 'Operational',
    operational: !building.isDestroyed,
    // passiveIncomeRate remains as a compatibility alias for older saves and
    // callers; EconomySystem uses the explicit gross/cost fields.
    passiveIncomeRate: grossIncomeRate,
    grossIncomeRate,
    operatingCostRate,
    happinessModifier,
    landValueModifier: finiteNumber(
      spec.landValueModifier,
      happinessModifier * 0.6 + (amenityRadius > 0 ? 3 : 0)
    ),
    position: getWorldPosition(building),
    amenityRadius,
    services: {
      power: {
        capacity: nonNegative(spec.powerSupply),
        reach: serviceReach(spec, 'power', nonNegative(spec.powerSupply)),
        demand: nonNegative(spec.powerDemand)
      },
      water: {
        capacity: nonNegative(spec.waterSupply),
        reach: serviceReach(spec, 'water', nonNegative(spec.waterSupply)),
        demand: nonNegative(spec.waterDemand)
      },
      fire: {
        capacity: nonNegative(spec.fireCoverage),
        reach: serviceReach(spec, 'fire', nonNegative(spec.fireCoverage)),
        demand: nonNegative(
          spec.fireDemand,
          Math.ceil((population + employees) / 180)
        )
      }
    }
  };
}

export default createBuildingEconomyRecord;

File: godot/MetroPulse.Domain/Economy/BuildingEconomyAdapter.cs
using MetroPulse.Domain.Content;
using MetroPulse.Domain.World;

namespace MetroPulse.Domain.Economy;

/// <summary>
/// The single mapping boundary from authored/catalog building data to the economy authority.
/// </summary>
public static class BuildingEconomyAdapter
{
    public static EconomyBuilding FromCatalog(
        BuildingDefinition spec,
        string id,
        EconomyPoint position,
        bool operational = true)
    {
        ArgumentNullException.ThrowIfNull(spec);
        if (string.IsNullOrWhiteSpace(id)) throw new ArgumentException("An economy building ID is required.", nameof(id));
        ArgumentNullException.ThrowIfNull(position);

        double authoredIncome = FiniteOrZero(spec.IncomePerMinute);
        double grossIncomeRate = Math.Max(0, authoredIncome) / 60;
        double operatingCostRate = Math.Max(0, -authoredIncome) / 60;
        int population = ToCount(spec.Residents);
        int employees = ToCount(spec.Employees);
        double amenityRadius = NonNegative(spec.AmenityRadius);
        double happiness = FiniteOrZero(spec.Happiness);

        return new EconomyBuilding(id.Trim(), grossIncomeRate, operatingCostRate, operational)
        {
            Name = spec.Name ?? id.Trim(),
            Kind = spec.Category ?? spec.GeneratorType ?? ConstructionCategories.Commercial,
            Value = NonNegative(spec.Value),
            Employees = employees,
            JobCapacity = employees,
            Population = population,
            HousingCapacity = population,
            Status = spec.Status ?? "Operational",
            PassiveIncomeRate = grossIncomeRate,
            HappinessModifier = happiness,
            LandValueModifier = happiness * 0.6 + (amenityRadius > 0 ? 3 : 0),
            Position = position,
            AmenityRadius = amenityRadius,
            Services = new EconomyBuildingServices
            {
                Power = new EconomyBuildingService(
                    NonNegative(spec.PowerSupply),
                    NonNegative(spec.PowerDemand),
                    ServiceReach(spec.PowerReach, spec.PowerSupply, 180)),
                Water = new EconomyBuildingService(
                    NonNegative(spec.WaterSupply),
                    NonNegative(spec.WaterDemand),
                    ServiceReach(spec.WaterReach, spec.WaterSupply, 160)),
                Fire = new EconomyBuildingService(
                    NonNegative(spec.FireCoverage),
                    Math.Ceiling((population + employees) / 180d),
                    ServiceReach(spec.FireReach, spec.FireCoverage, 140)),
            },
        };
    }

    public static EconomyBuilding FromAuthoredSkyline(WorldObjectDefinition definition, int ordinal)
    {
        ArgumentNullException.ThrowIfNull(definition);
        if (definition.ChunkId != "InitialSkyline" || definition.Kind != "building")
        {
            throw new ArgumentException("Only authored InitialSkyline building records can use the skyline adapter.", nameof(definition));
        }
        if (ordinal < 1) throw new ArgumentOutOfRangeException(nameof(ordinal));

        int employees = (int)Math.Round(definition.Size.Y * 8, MidpointRounding.AwayFromZero);
        double value = Math.Max(150_000, Math.Round((definition.Size.Y * 8_500) + (employees * 180)));
        return new EconomyBuilding($"existing-{ordinal}")
        {
            Name = definition.Id,
            Kind = ConstructionCategories.Commercial,
            Value = value,
            Employees = employees,
            JobCapacity = employees,
            Status = "Open",
            Position = new EconomyPoint(definition.Position.X, definition.Position.Z),
        };
    }

    public static IReadOnlyList<EconomyBuilding> AdaptAuthoredSkyline(MvpWorldLayout layout)
    {
        ArgumentNullException.ThrowIfNull(layout);
        EconomyBuilding[] records = layout.Objects
            .Where(item => item.ChunkId == "InitialSkyline" && item.Kind == "building")
            .Select((item, index) => FromAuthoredSkyline(item, index + 1))
            .ToArray();
        return Array.AsReadOnly(records);
    }

    private static int ToCount(double? value) =>
        (int)Math.Round(NonNegative(value), MidpointRounding.AwayFromZero);

    private static double FiniteOrZero(double? value) => value is double number && double.IsFinite(number) ? number : 0;

    private static double NonNegative(double? value) => Math.Max(0, FiniteOrZero(value));

    private static double ServiceReach(double? reach, double? capacity, double defaultReach)
    {
        double normalizedReach = NonNegative(reach);
        return normalizedReach > 0 || NonNegative(capacity) == 0 ? normalizedReach : defaultReach;
    }
}
